### ISSUE TITLE
stream_to seems to work correctly but no data appears in the table

### ISSUE THREAD

**Issue by mveygman** [jtv/libpqxx#408]
I am attempting to use the stream_to to write data to the table and running into several issues:

1.  Data seems to be written successfully but doesn't appear in the table:

test code:

```
#include <iostream>
#include <pqxx/pqxx>

int main() {
    pqxx::connection conn{<your favorite test URL>);
    pqxx::transaction txn(conn);
    pqxx::stream_to current_stream{txn, "test_table", std::vector<std::string>{"value"}};

    for(int idx = 0 ; idx < 1000 ; ++idx) {
        auto val = std::make_tuple(idx);
        current_stream << val;
   }
   current_stream.complete();
}
````

Table is created as: 
`
create table test_table(value integer);
`

The code, at least according to the documentation, should put values of 0 to 999 into the table `test_table` but as of now this is empty.

2.  If the table is created as: `create table schema.test_table(value integer);` version 7.3.1 of the library will blow up with error: "ERROR: relation \"schema.test_table\" doesn't exist.".

Reason for this is the following line: https://github.com/jtv/libpqxx/blob/master/src/stream_to.cxx#L31  It quote escapes the table name for the copy command whereas it doesn't do this if the fields are not specified.

I am using version `7.3.1` and `7.2.0` of the library.  Issue 2 is specific to version `7.3.1`

Postgresql version is 12.

**Comment by jtv** [comment]
Did you commit your transaction?

I'm guessing the problem with the schema is that the library puts quotes
around the whole "table name" string, but this includes the schema name!
That's a nasty one.  I think I'll have to scan for that dot, which requires
some knowledge about the encoding in use.

**Comment by mveygman** [comment]
No.  When I tried to do `txn.commit()` it threw a different error.  "Attempt to commit transaction with stream_to 'test_table' still open"

Is there a reason to quote escape the table?  It seems that normally creating the copy command doesn't require the table to be quote escaped.

**Comment by mveygman** [comment]
Actually the commit helps.  But it has to be done after the complete().  Which is not documented but seems to be an absolute requirement.

**Comment by jtv** [comment]
Yes, the stream stays open until you've completed it, and until then, you
can't commit.  I'll try to improve documentation but right now I'm having a
rather painful problem that stops me from doing any work at the computer.

The quoting is needed in case the name contains any special characters,
such as spaces; and it enforces consistently case-sensitive handling of the
names.

**Comment by mveygman** [comment]
Just did a quick test with copy command with postgres.  You will need to quote both schema_name and table name separately.  I don't know if it is possible to use . in the table name.

**Comment by Laro88** [comment]
I wrote a little gist a couple of hours ago @mveygman as I am also working with stream_to (a lot of MQTT payloads actually)
https://gist.github.com/Laro88/ec2cd952723bcbb829afc63658eb03e8

**Comment by jtv** [comment]
Meanwhile I'm pondering how to include the schema name.  I could add a
static factory function with a name like stream_to::schema_table(schema,
table, columns), and separately, a stream_to::table(table, columns) without
the table name.

Surprisingly, the problem with that solution (and I had been stuck on this
for some time) is how I provide the options of either a static or a dynamic
list of columns!

In most situations, I think it would be nice if you could write a stream of
3 columns like...

    auto s = stream_to::table(vehicle, type, licence_plate, owner);

But sometimes of course people will want a _dynamic_ list of columns.  And
in the presence of all the different string types it can be hard to detect
the difference between an argument that's a single column name, and one
that's a container of column names - at least until we can use C++20.

*So...*  Does anyone have any good ideas for naming a parameter pack of
strings, as opposed to a container / sequence / view / span of strings?

**Comment by KayEss** [comment]
I don't know if it's a good idea, but something like this *should* work:

    auto stream_to::table(std::string_view table_name, ....);
    auto stream_to::talbe(std::pair<std::string_view, std::string_view> schema_table_name, ...);
    
Then use as:

    stream.table("mytable", ...);

or

    stream.table({"myschema", "mytable"}, ...);

**Comment by jtv** [comment]
Nice one, @Kayess!

**Comment by jtv** [comment]
@mveygman @Laro88 @KayEss I just pushed a big change to how you create a `stream_from` or a `stream_to`.  No more overloaded constructors!  Instead you get factories like...

    // extern pqxx::work tx;
    auto t1 = pqxx::stream_to::table(tx, {"mytable"}, {"column1", "column2"});
    auto t2 = pqxx::stream_to::table(tx, {"myschema", "mytable"}, {"col1"});

    // These particular names don't actually any quoting or escaping, so this
    // is safe...  and slightly more efficient.
    auto t3 = pqxx::stream_to::raw_table(tx, "myschema.mytable", "col1, col2, col3");

    // These table and column names may need quoting, so do that too.
    // Use pqxx::connection::quote_table() and pqxx::connection::quote_columns().
    // The latter can handle either a static `std::initializer_list<std::string_view>` or
    // anything that iterates things that we can convert to `std::string_view`.
    // Can't wait for C++20 as our baseline!

    // extern string schema, table;
    // extern std::vector<std::string> columns;
    auto const &conn = tx.conn();
    auto t4 = pqxx::stream_to::raw_table(
        tx,
        conn.quote_table({schema, table}),
        conn.quote_columns(columns));

And similar for `stream_from`, which in addition, already had a factory called `pqxx::stream_from::query()`.

There's also a new convenience typedef called `pqxx::table_path`, which is an alias for `std::initializer_list<std::string_view>`.  (Thanks @KayEss for that tip!)  A `table_path` may contain just a table name; or a schema name and a table name; or a database name and a schema name and a table name.  Quoting e.g. `{"My DB", "My Schema", "My Table"}` will produce `"My DB"."My Schema"."My Table".

Anyone who's interested, please try this out and let me know anything you'd like to see changed before it goes into a 7.4 release!

**Comment by jtv** [comment]
@mveygman @Laro88 @KayEss could you give the new factory functions in master a go?  Have a look at `from_stream::table()`, `to_stream::table`, `from_stream::raw_table()`, and `to_stream::raw_table()`.

**Comment by jtv** [comment]
No feedback so far.  I may just release them without prior feedback.

**Comment by Laro88** [comment]
It looks good and straightforward @jtv, haven’t done any testing though. I am using vcpkg with a fixed git branch so I am always training a bit behind. If you need testing with latest master then let me know and I will find a clean machine.

**Comment by mveygman** [comment]
@jtv I looked through the code and it looks fine.  I didn't get a chance to run some formal tests on it.

**Comment by jtv** [comment]
Thanks @Laro88 & @mveygman!  It's no big deal code-wise, really, but setting a new direction in API styles can always do with a bit of back-and-forth from different perspectives.

**Comment by jtv** [comment]
Just released 7.4.0, with the new constructors and better support for schemata¹ using a new type alias.

I think that solves this problem, so I'll close the bug.

—
¹) Technically I think that's the correct Greek plural of "schema."  Sometimes I like to be a pedant.

**Comment by hharng** [comment]
{"schema","table_name"} worked for me.
Thank you!
